src/proto.rs

```diff
@@ -97,26 +106,207 @@ impl MessageHeader {
+        writer.flush().await
+    }
+
+    /// Decodes a MessageHeader from reader.
+    pub async fn decode_from(
+        mut reader: impl tokio::io::AsyncReadExt + Unpin,
+    ) -> std::io::Result<MessageHeader> {
+        let mut content = vec![0; MESSAGE_HEADER_LENGTH];
+        reader.read_exact(&mut content).await?;
+        Ok(MessageHeader::from(&content))
+    }
+}
+
+/// Generic message of ttrpc.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct GenMessage {
+    pub header: MessageHeader,
+    pub payload: Vec<u8>,
+}
+
+#[cfg(feature = "async")]
+impl GenMessage {
+    /// Encodes a MessageHeader to writer.
+    pub async fn encode_to(
+        &self,
+        mut writer: impl tokio::io::AsyncWriteExt + Unpin,
+    ) -> TtResult<()> {
+        self.header
+            .encode_to(&mut writer)
+            .await
+            .map_err(|e| Error::Socket(e.to_string()))?;
+        writer
+            .write_all(&self.payload)
+            .await
+            .map_err(|e| Error::Socket(e.to_string()))?;
+        Ok(())
+    }
+
+    /// Decodes a MessageHeader from reader.
+    pub async fn decode_from(mut reader: impl tokio::io::AsyncReadExt + Unpin) -> TtResult<Self> {
+        let header = MessageHeader::decode_from(&mut reader)
+            .await
+            .map_err(|e| Error::Socket(e.to_string()))?;
+
+        if header.length > MESSAGE_LENGTH_MAX as u32 {
+            return Err(get_rpc_status(
+                Code::INVALID_ARGUMENT,
+                format!(
+                    "message length {} exceed maximum message size of {}",
+                    header.length, MESSAGE_LENGTH_MAX
+                ),
+            ));
+        }
+
+        let mut content = vec![0; header.length as usize];
+        reader
+            .read_exact(&mut content)
+            .await
+            .map_err(|e| Error::Socket(e.to_string()))?;
+
+        Ok(Self {
+            header,
+            payload: content,
+        })
+    }
+}
+
+/// TTRPC codec, only protobuf is supported.
+pub trait Codec {
+    type E;
+
+    fn size(&self) -> u32;
+    fn encode(&self) -> Result<Vec<u8>, Self::E>;
+    fn decode(buf: impl AsRef<[u8]>) -> Result<Self, Self::E>
+    where
+        Self: Sized;
+}
+
+impl<M: protobuf::Message> Codec for M {
+    type E = protobuf::error::ProtobufError;
+
+    fn size(&self) -> u32 {
+        self.compute_size()
+    }
+
+    fn encode(&self) -> Result<Vec<u8>, Self::E> {
+        let mut buf = vec![0; self.compute_size() as usize];
+        let mut s = CodedOutputStream::bytes(&mut buf);
+        self.write_to(&mut s)?;
+        s.flush()?;
+        Ok(buf)
+    }
+
+    fn decode(buf: impl AsRef<[u8]>) -> Result<Self, Self::E> {
+        let mut s = CodedInputStream::from_bytes(buf.as_ref());
+        M::parse_from(&mut s)
+    }
+}
+
+/// Message of ttrpc.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct Message<C> {
+    pub header: MessageHeader,
+    pub payload: C,
+}
```

Comment:
`Message<C>` will take a generic parameter, which is a problem when using `mpsc::channel`, because a channel can only send one type of message, so GenMessage is needed.

Reply:
Currently, `GenMessage` has `encode_to, decode_from` and `Message<C>` also have this 2 function, how about `type `GenMessage = Message<Vec<u8>>`.